presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SingleDistinctOptimizer.java

```diff
@@ -0,0 +1,150 @@
+                                                            false,
+                                                            aggregationFunctionCall.getArguments());
+                aggregations.put(entry.getKey(), functionCall);
+            }
+
+            return new AggregationNode(idAllocator.getNextId(),
+                                        source,
+                                        node.getGroupBy(),
+                                        aggregations,
+                                        node.getFunctions(),
+                                        Collections.emptyMap(),
+                                        node.getStep(),
+                                        node.getSampleWeight(),
+                                        node.getConfidence(),
+                                        node.getHashSymbol());
+        }
+
+        boolean canOptimize(AggregationNode node)
+        {
+            // all aggregation functions have a single common distinct mask symbol
+            // AND all aggregation functions have mask
+            return node.getMasks().values().stream().distinct().count() == 1 &&
+                    node.getMasks().size() == node.getAggregations().size();
+        }
+
+        public PlanNode visitMarkDistinct(MarkDistinctNode node, RewriteContext<Optional<Symbol>> context)
+        {
+            Optional<Symbol> mask = context.get();
+            if (mask.isPresent() && mask.get().equals(node.getMarkerSymbol())) {
+                // rewrite Distinct into GroupBy
+                AggregationNode aggregationNode = new AggregationNode(idAllocator.getNextId(),
+                                                                        node.getSource(),
+                                                                        node.getDistinctSymbols(),
+                                                                        Collections.emptyMap(),
+                                                                        Collections.emptyMap(),
+                                                                        Collections.emptyMap(),
+                                                                        SINGLE,
+                                                                        Optional.empty(),
+                                                                        1.0,
+                                                                        node.getHashSymbol());
+
+                ImmutableMap.Builder<Symbol, Expression> outputSymbols = ImmutableMap.builder();
+                for (Symbol symbol : node.getDistinctSymbols()) {
+                    Expression expression = new QualifiedNameReference(symbol.toQualifiedName());
+                    outputSymbols.put(symbol, expression);
+                }
+
+                if (aggregationNode.getHashSymbol().isPresent()) {
+                    Expression hashExpression = getHashExpression(node.getDistinctSymbols());
```

Comment:
This doesn't look right. If the MarkDistinctNode has a "hashSymbol", it means that the hash is pre-computed by its child. You should be able to project it as-is.